### PR TITLE
[TRTLLM-6091][docs] Update docs/trtllm sampler 1.0

### PR DESCRIPTION
**Pull request by stnie**
## Description

Added Documentation for TRTLLM Sampling. Reused the old Documentation and added a new section regarding Beam Search

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: docs/source/1.0/features/sampling.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+To enable beam search, you must:
+
+1. Enable the `use_beam_search` option in the `SamplingParams` object
+2. Set the `max_beam_width` parameter in the `LLM` class to match the `best_of` parameter in `SamplingParams`
+3. Disable overlap scheduling using the `disable_overlap_scheduler` parameter of the `LLM` class
+
+Parameter Configuration:
+- `best_of`: Controls the number of beams processed during generation (beam width)
+- `n`: Controls the number of output sequences returned (can be less than `best_of`)
+- If `best_of` is omitted, it will be implicitly set to `n`
```
Comment (chenopis) [inline docs/source/1.0/features/sampling.md:36]: Phrase it as a default, i.e.

- If `best_of` is omitted, the number of beams processed defaults to `n`

**Review by chenopis — APPROVED** [review]
Some small suggested edits.

File: docs/source/1.0/features/sampling.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Sampling
+The PyTorch backend supports most of the sampling features that are supported on the C++ backend, such as temperature, top-k and top-p sampling, beam search, stop words, bad words, penalty, context and generation logits, and log probs.
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I updated the test passage

File: docs/source/1.0/features/sampling.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+## General usage
+
+In order to use this feature, it is necessary to enable option `enable_trtllm_sampler` in the `LLM` class, and pass a `SamplingParams` object with the desired options as well. The following example prepares two identical prompts which will give different results due to the sampling parameters chosen:
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I updated the test passage

File: docs/source/1.0/features/sampling.md (hunk 1 above)
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I updated the test passage

File: docs/source/1.0/features/sampling.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+sampling_params = SamplingParams(
+        temperature=1.0,
+        top_k=8,
+        top_p=0.5,
+    )
+llm.generate(["Hello, my name is",
+            "Hello, my name is"], sampling_params)
+```
+
+When using speculative decoders such as MTP or Eagle-3, the `enable_trtllm_sampler` option is not yet supported and therefore the subset of sampling options available is more restricted.
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I updated the test passage

**Review by chenopis — APPROVED** [review]
LGTM

**Review by chenopis — APPROVED** [review]
LGTM

**Review by syuoni — APPROVED** [review]

**Review by syuoni — CHANGES_REQUESTED** [review]
Hi @stnie ,

Sorry, I revisited the [doc plan](https://docs.google.com/spreadsheets/d/1PfY5M87TanVUf9hbUwkElQFr7M8J_kw_WqMoYNumiIU/edit?gid=0#gid=0) 33rd row: Sampling. It should cover greedy, beam, logit processor, guided decoding, xgrammar, etc.

Seems the logits processor, guided decoding (i.e., xgrammar/llguidance) are missing from the PR. Could you please add them? Thanks

File: docs/source/1.0/features/sampling.md [hunk 5]
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
...
+llm.generate(["Hello, my name is",
+            "Hello, my name is"], sampling_params)
+```
+
+## Guided decoding
+
+Guided decoding controls the generation outputs to conform to pre-defined structured formats, ensuring outputs follow specific schemas or patterns.
+
+The PyTorch backend supports guided decoding with the [XGrammar](https://github.com/mlc-ai/xgrammar) backend and the following formats:
+- JSON
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I added both options to the list

File: docs/source/1.0/features/sampling.md [hunk 6]
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
...
+- Regular expressions
+- Extended Backus-Naur form (EBNF) grammar
+- Structural tags
+
+To enable guided decoding, you must:
+
+1. Set the `guided_decoding_backend` parameter to `'xgrammar'` in the `LLM` class
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: I added it onto the general description as well as the step-by-step list

File: docs/source/1.0/features/sampling.md [hunk 7]
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
...
+
+You can find a more detailed example on guided decoding [here](../../../../examples/llm-api/llm_guided_decoding.py).
+
+## Logits processor
+
+Logits processors allow you to modify the logits produced by the network before sampling, enabling custom generation behavior and constraints.
+
+The PyTorch backend supports two types of logits processing:
+- **Per-prompt processing**: Processes logits for each prompt individually, providing fine-grained control
+- **Batched processing**: Processes logits for the entire batch together, optimizing performance for larger batch sizes
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: removed

File: docs/source/1.0/features/sampling.md [hunk 8]
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
...
+        return logits
+
+llm = LLM(model='nvidia/Llama-3.1-8B-Instruct-FP8')
+sampling_params = SamplingParams(
+        logits_processor=MyCustomLogitsProcessor()
+    )
+llm.generate(["Hello, my name is"], sampling_params)
+```
+
+### Batched logits processing
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: removed

File: docs/source/1.0/features/sampling.md [hunk 9]
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
...
+The PyTorch backend supports guided decoding with the [XGrammar](https://github.com/mlc-ai/xgrammar) backend and the following formats:
+- JSON
+- Regular expressions
+- Extended Backus-Naur form (EBNF) grammar
+- Structural tags
+
+To enable guided decoding, you must:
+
+1. Set the `guided_decoding_backend` parameter to `'xgrammar'` in the `LLM` class
+2. Disable overlap scheduling using the `disable_overlap_scheduler` parameter of the `LLM` class
```
[the remark this answers is not in the document]
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: adjusted the examples and the step-by-step list to no longer disable overlap scheduler

File: docs/source/1.0/features/sampling.md [hunk 10]
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
...
+class MyCustomLogitsProcessor(LogitsProcessor):
+    def __call__(self,
+        req_id: int,
+        logits: torch.Tensor,
+        token_ids: List[List[int]],
+        stream_ptr: Optional[int],
+        client_id: Optional[int]
+    ) -> None:
+        # Implement your custom inplace logits processing logic
+        logits = 2 * logits
```
Comment (syuoni) [inline docs/source/1.0/features/sampling.md:126]: Must use in-place op so that the logits processing is effective.

```suggestion
        logits *= logits
```

File: docs/source/1.0/features/sampling.md [hunk 11]
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
...
+- Structural tags
+
+To enable guided decoding, you must:
+
+1. Set the `guided_decoding_backend` parameter to `'xgrammar'` or `'llguidance'` in the `LLM` class
+2. Create a [`GuidedDecodingParams`](../../../../tensorrt_llm/sampling_params.py#L14) object with the desired format specification
+    * Note: Depending on the type of format, a different parameter needs to be chosen to construct the object (`json`, `regex`, `grammar`, `strucutral_tag`).
+3. Pass the `GuidedDecodingParams` object to the `guided_decoding` parameter of the `SamplingParams` object
+
+The following example demonstrates guided decoding with JSON format:
```
Comment (syuoni) [inline docs/source/1.0/features/sampling.md:81]:
```suggestion
The following example demonstrates guided decoding with a JSON schema:
```

File: docs/source/1.0/features/sampling.md [hunk 12]
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
...
+from tensorrt_llm.llmapi import GuidedDecodingParams
+
+llm = LLM(model='nvidia/Llama-3.1-8B-Instruct-FP8',
+          guided_decoding_backend='xgrammar')
+structure = '{"title": "Example JSON", "type": "object", "properties": {...}}'
+guided_decoding_params = GuidedDecodingParams(json=structure)
+sampling_params = SamplingParams(
+        guided_decoding=guided_decoding_params,
+    )
+llm.generate(["Generate a JSON response"], sampling_params)
```
Comment (syuoni) [inline docs/source/1.0/features/sampling.md:94]:
```suggestion
llm.generate("Generate a JSON response", sampling_params)
```

**Review by syuoni — COMMENTED** [review]
LGTM, thanks!

**Review by syuoni — APPROVED** [review]

File: docs/source/1.0/features/sampling.md (hunk 11 above)
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: done

File: docs/source/1.0/features/sampling.md (hunk 10 above)
Comment (stnie) [reply docs/source/1.0/features/sampling.md]: done
